tests/unit/s2n_async_generic_cb_test.c

```diff
@@ -0,0 +1,221 @@
+        },
+        {
+            .allow_list = S2N_ASYNC_VERIFY,     /* ASYNC_VERIFY is allowed. */
+            .cb_return = S2N_SUCCESS,   /* Client auth is enabled for all the sync tests. In this case, */
+            .cb_invoked = 2,            /* a successfule handshake performs pkey_verify() twice. */
+        },
+        {
+            .allow_list = ALLOW_ALL,    /* ASYNC_VERIFY is allowed. */
+            .cb_return = S2N_FAILURE,   /* Handshake failed because the generic */
+            .cb_invoked = 1,            /* callback failed in the first attempt. */
+        },
+    };
+    /* clang-format on */
+
+    const char *versions[] = { "default", "default_tls13" };
+
+    /* Sync Test: 1) ASYNC_VERIFY is not allowed, or 2) op_perform() invoked in the callback. */
+    for (int test_idx = 0; test_idx < s2n_array_len(verify_sync); test_idx++) {
+        for (int version_idx = 0; version_idx < s2n_array_len(versions); version_idx++) {
+            DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(), s2n_config_ptr_free);
+            EXPECT_NOT_NULL(config);
+            EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+            EXPECT_SUCCESS(s2n_config_set_verification_ca_location(config, S2N_DEFAULT_TEST_CERT_CHAIN, NULL));
+            EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, versions[version_idx]));
+            EXPECT_SUCCESS(s2n_config_set_client_auth_type(config, S2N_CERT_AUTH_REQUIRED));
+
+            struct s2n_async_generic_cb_test data = { .invoke_perform = true, .result = verify_sync[test_idx].cb_return };
+            EXPECT_SUCCESS(s2n_config_set_async_generic_callback(config, async_op_perform_in_callback, &data));
+            EXPECT_SUCCESS(s2n_config_set_async_op_allow_list(config, verify_sync[test_idx].allow_list));
+
+            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server_conn);
+            EXPECT_SUCCESS(s2n_connection_set_config(server_conn, config));
+
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client_conn);
+            EXPECT_SUCCESS(s2n_connection_set_config(client_conn, config));
+            EXPECT_SUCCESS(s2n_connection_set_blinding(client_conn, S2N_SELF_SERVICE_BLINDING));
+            EXPECT_SUCCESS(s2n_set_server_name(client_conn, "localhost"));
+
+            DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
+            EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+            EXPECT_SUCCESS(s2n_connection_set_io_pair(client_conn, &io_pair));
+            EXPECT_SUCCESS(s2n_connection_set_io_pair(server_conn, &io_pair));
+
+            bool verify_allowed = s2n_async_is_op_in_allow_list(config, S2N_ASYNC_VERIFY);
```

Comment:
It seems better to branch based on the test case and then assert on `s2n_async_is_op_in_allow_list`. Otherwise if `s2n_async_is_op_in_allow_list` gives the wrong value, you might not be testing what you think you are.